### PR TITLE
non-cryptographically secure but likely good enough prng

### PR DESCRIPTION
**Pull request by calvinmetcalf**
for use in older browsers for testing and other such

### CODE REVIEW

**Comment by dcousens** [comment]
Any chance of making `chacha20` a module on its own instead?

**Comment by calvinmetcalf** [comment]
it is it's own module, that module just assumes you actually want to use it
correctly instead of using all 512 bits of state like this does, not that
this really needs to

**Comment by dcousens** [comment]
> non-cryptographically secure

I feel like that isn't justification for this to be acceptable.
I don't know [or have the time to currently] understand the implications of using ChaCha20 in this way.

I can't reasonably ACK this,  as it just doesn't feel like it is worthwhile?

File: prng.js [hunk 1]
```diff
@@ -1,25 +1,21 @@
...
+    .digest();
+  this.chacha = new ChaCha(blendedSeed);
 };
-
-PRNG.prototype.checkReseed = function (len) {
-  if ((this.bytes + len) >= maxInt) {
-    this.seed();
-  }
+PRNG.prototype._seed = function () {
+  throw new Error('you must impliment me');
```
Comment (dcousens) [inline prng.js:15]: `'Unimplemented'`

File: package.json [hunk 2]
```diff
@@ -11,6 +11,8 @@
   "license": "MIT",
   "dependencies": {
     "chacha": "^1.1.0",
```
Comment (dcousens) [inline package.json]: `chacha` no longer being used

**Comment by dcousens** [comment]
What uses this anyway?

**Comment by calvinmetcalf** [comment]
The idea for this was for the the tests to be run in Opera and older IE,
it's based on the prng in rust.

**Comment by dcousens** [comment]
> it's based on the prng in rust.

Source?

**Comment by dcousens** [comment]
> The idea for this was for the the tests to be run in Opera and older IE

We abandoned those because they didn't support the API anyway?

**Comment by dcousens** [comment]
By the way,  I don't mean to put you through this much scrutiny every time,  I'll buy you a beer one day :+1: .
Just doing due diligence since its mostly only the two of us keeping these repositories in check.

**Comment by calvinmetcalf** [comment]
First to make it clear, this isn't something I intend to try to make a part of crypto browserify that is distributed as part of browserify. With old ie and Opera we can't test the hashes and ciphers that should work there because they don't support cryptographically secure random number generators, but all we are needing in the tests are statistically random numbers. 

And I have no problem with scrutiny for stuff in this org, it's kind of the point. (this is the [rust rng](https://doc.rust-lang.org/rand/rand/chacha/struct.ChaChaRng.html) I was inspired by).

**Comment by dcousens** [comment]
> we can't test the hashes and ciphers that should work there because they don't support cryptographically secure random number generators

Which tests need a PRNG?

**Comment by calvinmetcalf** [comment]
ciphers rely on random data

**Comment by dcousens** [comment]
@calvinmetcalf link?

**Comment by calvinmetcalf** [comment]
https://github.com/crypto-browserify/crypto-browserify/blob/master/test/aes.js

**Comment by dcousens** [comment]
I don't understand why that needs to be random?
Aren't the point of these tests to be deterministic?

Why not just generate 1-2MB of vectors and put them in a repo and test against that, you'll receive enough of a sample that you'd likely see the same result in the wild running as though you were running them randomly each time.

**Comment by calvinmetcalf** [comment]
That's what we do in the repo, this is intended as a less comprehensive
test to prevent accidental regressions and statistically random inputs does
that perfectly

**Comment by dcousens** [comment]
> prevent accidental regressions and statistically random inputs does that perfectly

That just sounds like you need more vectors.
TRAVIS doesn't run often enough that you'd be getting much statistically random variation...

**Comment by calvinmetcalf** [comment]
It's mainly to prevent something being wrong with the require or how it's
brought into crypto-browserify, I considered random data a compromise
between just a the same string for each test and a large number of vector
tests duplicated from the module

**Comment by dominictarr** [comment]
There have been bugs found previously where the static values passed, but someone discovered a bug that happened with other values - so we randomly choose the values (key, etc) to get fairer coverage.

It just needs to be a fair distribution, but not cryptographically secure randomness.

**Comment by dcousens** [comment]
> There have been bugs found previously where the static values passed, but someone discovered a bug that happened with other values - so we randomly choose the values (key, etc) to get fairer coverage.

I'm still not understanding how you couldn't just pre-generate a uniform distribution, and use that.
If the distribution isn't used for actual cryptographic purposes,  what use is using an RNG in this case?
You're better off having deterministic tests.
IMHO.

I'm happy to merge this if we put a bigger warning on the repository,  otherwise I'll have to do an in depth review which might be a few days until I have time.

**Comment by dominictarr** [comment]
it's just how the tests where originally written... a static test vector would be fine.

**Comment by dominictarr** [comment]
we do have static vectors in some cases where there are appropiate standards.

**Comment by calvinmetcalf** [comment]
how about this

**Comment by dcousens** [comment]
I'd still like to remove the need for this repository if possible.

**Pull request merged**